### PR TITLE
add resampling code

### LINKED ISSUE #25: Add optional rubato dependency
So we can work with PCM audio that isn't 8kHz or 16kHz.

### PR DESCRIPTION
**Pull request by WenqingZong**
Fix #25 

Now `process` method can resample any received pcm audio to 16000Hz in `auto-resample` feature.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -6,21 +6,22 @@ readme = "README.md"
 license = "MIT"
 edition = "2021"
 
+[features]
+default = ["static-model", "auto_resample"]
+static-model = []
+auto_resample = ["dep:rubato"]
```
Comment (xd009642) [inline Cargo.toml:12]: Name quibble I'd prefer something like audio_resampler or just keeping it as the rubato feature and explaining it in the readme

File: Cargo.toml [hunk 2]
```diff
@@ -6,21 +6,22 @@ readme = "README.md"
...
+serde = ["dep:serde"]
+
 [dependencies]
 anyhow = "1.0.86"
+hound = "3.5.1"
```
Comment (xd009642) [inline Cargo.toml:17]: I wouldn't add any audio file processing to this project. Have the users do that themselves

File: src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+use rubato::{
+    Resampler, SincFixedIn, SincInterpolationParameters, SincInterpolationType, WindowFunction,
```
Comment (xd009642) [inline src/utils.rs:2]: I think this was discussed on slack different resamplers have different aliasing characteristics. If you can make sure you pick the most accurate one it'll avoid potential issues at start/end of frame

File: src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+};
+use tracing::info;
```
Comment (xd009642) [inline src/utils.rs:4]: Trace or debug logs we don't want to be too noisy to users

File: tests/audio/sample_1.mp3
Comment (xd009642) [inline tests/audio/sample_1.mp3:1]: Is this MP3 used anywhere?

**Review by xd009642 — CHANGES_REQUESTED** [review]

**Comment by xd009642** [comment]
A few review comments

File: src/lib.rs [hunk 5]
```diff
@@ -1,4 +1,8 @@
 #![doc = include_str!("../README.md")]
+
+pub mod utils;
```
Comment (xd009642) [inline src/lib.rs:3]: The module above the import statements is a bit weird formatting wise, add a line break or two and put it below the imports (that's more idiomatic)

File: src/lib.rs [hunk 6]
```diff
@@ -113,7 +117,18 @@ impl VadSession {
 
     /// Pass in some audio to the VAD and return a list of any speech transitions that happened
     /// during the segment.
+    ///
+    /// If "auto_resample" feature is enabled, then this function will resample what ever it
```
Comment (xd009642) [inline src/lib.rs:121]: Update feature name in doc comment. "what ever" should be "whatever". And line below change "then user input is expected to be of the right" to "user input is expected to be either" and maybe mention the unit is hertz/Hz

File: src/lib.rs [hunk 7]
```diff
@@ -113,7 +117,18 @@ impl VadSession {
...
+    /// receives to 16000Hz samples. If not enabled, then user input is expected to be of the right
+    /// sample rate: 8000 or 16000.
     pub fn process(&mut self, audio_frame: &[f32]) -> Result<Vec<VadTransition>> {
+        #[cfg(feature = "audio_resampler")]
+        let audio_frame = if self.config.sample_rate != 16000 {
```
Comment (xd009642) [inline src/lib.rs:126]: We don't need to resample 8000Hz audio to 16000Hz

File: src/lib.rs [hunk 8]
```diff
@@ -359,6 +374,42 @@ impl Default for VadConfig {
...
+        }
+    }
+
+    pub fn validate_config(&self) -> Result<()> {
+        #[cfg(feature = "audio_resampler")]
+        return Ok(());
+
+        #[cfg(not(feature = "audio_resampler"))]
+        if ![8000, 16000].contains(&self.sample_rate) {
+            bail!("You can only choose 8000 or 16000Hz sample rate")
```
Comment (xd009642) [inline src/lib.rs:406]: `Invalid sample rate of {}, expected either 8000Hz or 16000Hz` I think is a better message

File: tests/snapshot.rs [hunk 9]
```diff
@@ -218,7 +218,7 @@ fn get_audios() -> Vec<PathBuf> {
     for entry in fs::read_dir(&audio_dir).unwrap() {
         let entry = entry.unwrap();
         let path = entry.path();
-        if !path.is_dir() {
```
Comment (xd009642) [inline tests/snapshot.rs:221]: Do you still need this change now the mp3 is gone?

**Review by xd009642 — CHANGES_REQUESTED** [review]
A few more comments, mostly small tweaks

File: tests/snapshot.rs (hunk 9 above)
Comment (WenqingZong) [reply tests/snapshot.rs:221]: I'm planning to add it back in a follow up PR where I'll add audio file reading code so let's just keep it there.

File: .github/workflows/rust_ci.yml [hunk 10]
```diff
@@ -28,7 +28,7 @@ jobs:
         run: cargo test --all-features
 
       - name: Cargo build
-        run: cargo build
+        run: cargo build --all-features
```
Comment (xd009642) [inline .github/workflows/rust_ci.yml:31]: So we also want to do a check that we can't have a feature combination that causes build/check to fail (which can sometimes happen as you add more features) Given the simplicity of the project so far we could probably just get away with  a `cargo build --no-default-features` as serde missing won't break anything

**Review by xd009642 — CHANGES_REQUESTED** [review]
Noticed something else :see_no_evil: . I'll start looking at the example code now but just figured I'd leave this one as it's potentially a biggy

File: example/file_inference/src/audio_reading.rs [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    // Now we have interleaved audio, need to convert it to channel based.
+    let stereo_audio = interleaved_to_channel(interleave_samples, num_of_channels);
+
+    // Silero VAD only work with mono audio.
+    let mono_audio = stereo_to_mono(stereo_audio, multi_channel_strategy);
+
+    // Silero VAD only work with 8000 or 16000 Hz audio.
+    resample_pcm(
```
Comment (xd009642) [inline example/file_inference/src/audio_reading.rs:145]: Wouldn't it be nicer to just let the resampling happen inside the `VadSession` like the code allows? This example might make people think they have to resample and pass in resampled audio.

**Review by xd009642 — CHANGES_REQUESTED** [review]
And one final thing :sweat_smile:

File: example/file_inference/src/audio_reading.rs (hunk 11 above)
Comment (WenqingZong) [reply example/file_inference/src/audio_reading.rs]: done

**Review by xd009642 — APPROVED** [review]

**Pull request merged**